core/trino-main/src/main/java/io/trino/execution/SetTimeZoneTask.java

```diff
@@ -0,0 +1,114 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.slice.Slice;
+import io.trino.execution.warnings.WarningCollector;
+import io.trino.metadata.Metadata;
+import io.trino.security.AccessControl;
+import io.trino.spi.TrinoException;
+import io.trino.sql.planner.ExpressionInterpreter;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.IntervalLiteral;
+import io.trino.sql.tree.SetTimeZone;
+import io.trino.sql.tree.StringLiteral;
+import io.trino.transaction.TransactionManager;
+
+import javax.inject.Inject;
+
+import java.util.List;
+import java.util.TimeZone;
+
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static io.trino.SystemSessionProperties.TIME_ZONE_ID;
+import static io.trino.spi.StandardErrorCode.INVALID_LITERAL;
+import static io.trino.spi.type.TimeZoneKey.getTimeZoneKey;
+import static io.trino.spi.type.TimeZoneKey.getTimeZoneKeyForOffset;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static io.trino.type.IntervalDayTimeType.INTERVAL_DAY_TIME;
+import static io.trino.util.Failures.checkCondition;
+import static java.lang.String.format;
+
+public class SetTimeZoneTask
+        implements DataDefinitionTask<SetTimeZone>
+{
+    @Inject
+    public SetTimeZoneTask()
+    {
+    }
+
+    @Override
+    public String getName()
+    {
+        return "SET TIME ZONE";
+    }
+
+    @Override
+    public ListenableFuture<?> execute(
+            SetTimeZone statement,
+            TransactionManager transactionManager,
+            Metadata metadata,
+            AccessControl accessControl,
+            QueryStateMachine stateMachine,
+            List<Expression> parameters,
+            WarningCollector warningCollector)
+    {
+        String timeZoneId = statement.getTimeZone()
+                .map(timeZone -> getTimeZoneId(timeZone, metadata, accessControl, stateMachine))
+                .orElse(TimeZone.getDefault().getID());
+        stateMachine.addSetSessionProperties(TIME_ZONE_ID, timeZoneId);
+
+        return immediateFuture(null);
+    }
+
+    private static String getTimeZoneId(Expression timeZone, Metadata metadata, AccessControl accessControl, QueryStateMachine stateMachine)
+    {
+        if (timeZone instanceof IntervalLiteral) {
+            return intervalToTimeZoneId((IntervalLiteral) timeZone, metadata, accessControl, stateMachine);
+        }
```

Comment:
There's no requirement for the argument to be an interval literal. It only needs to be an expression of *type* interval (i.e., if could be a function call or other complex expression that returns such type)

Reply:
Thanks for the heads'up.
I made the appropriate modifications to support expressions given to `SET TIME ZONE`.

```
trino> set time zone concat_ws('/', 'America', 'Los_Angeles');
SET TIME ZONE
trino> select current_timestamp;
                    _col0                    
---------------------------------------------
 2021-06-08 01:31:18.761 America/Los_Angeles 
(1 row)
```

```
trino> set time zone fdsafdsa;
Query 20210608_092131_00001_r8ubx failed: line 1:15: Column 'fdsafdsa' cannot be resolved
set time zone fdsafdsa
```

```
trino> set time zone position('o' in 'Los Angeles');
Query 20210608_092540_00004_r8ubx failed: Time Zone expression "strpos('Los Angeles', 'o')" not supported
```

```
trino> SET TIME ZONE parse_duration('8h');
SET TIME ZONE
trino> SET TIME ZONE parse_duration('3601s');
Query 20210608_113400_00004_vfuxk failed: Invalid time zone offset interval: interval contains seconds
```


I've found that some of the exception messages are now misleading (due to the fact that the `expression` newly instroduced in the grammar of `SET TIME ZONE` statement are actually mainly used for `SELECT` statements).